cmd/erasure.go

```diff
@@ -18,105 +18,221 @@ package cmd
+	getLockers func() []dsync.NetLocker
+
+	// Locker mutex map.
+	nsMutex *nsLockMap
+
+	// Byte pools used for temporary i/o buffers.
+	bp *bpool.BytePoolCap
+
+	// TODO: ListObjects pool management, should be removed in future.
+	listPool *TreeWalkPool
+
+	mrfUploadCh chan partialUpload
+}
+
+// NewNSLock - initialize a new namespace RWLocker instance.
+func (xl xlObjects) NewNSLock(ctx context.Context, bucket string, object string) RWLocker {
+	return xl.nsMutex.NewNSLock(ctx, xl.getLockers, bucket, object)
 }
 
-// EncodeData encodes the given data and returns the erasure-coded data.
-// It returns an error if the erasure coding failed.
-func (e *Erasure) EncodeData(ctx context.Context, data []byte) ([][]byte, error) {
-	if len(data) == 0 {
-		return make([][]byte, e.dataBlocks+e.parityBlocks), nil
+// Shutdown function for object storage interface.
+func (xl xlObjects) Shutdown(ctx context.Context) error {
+	// Add any object layer shutdown activities here.
+	closeStorageDisks(xl.getDisks())
+	closeLockers(xl.getLockers())
+	return nil
+}
+
+// byDiskTotal is a collection satisfying sort.Interface.
+type byDiskTotal []DiskInfo
+
+func (d byDiskTotal) Len() int      { return len(d) }
+func (d byDiskTotal) Swap(i, j int) { d[i], d[j] = d[j], d[i] }
+func (d byDiskTotal) Less(i, j int) bool {
+	return d[i].Total < d[j].Total
+}
+
+// getDisksInfo - fetch disks info across all other storage API.
+func getDisksInfo(disks []StorageAPI) (disksInfo []DiskInfo, onlineDisks, offlineDisks madmin.BackendDisks) {
+	disksInfo = make([]DiskInfo, len(disks))
+
+	g := errgroup.WithNErrs(len(disks))
+	for index := range disks {
+		index := index
+		g.Go(func() error {
+			if disks[index] == nil {
+				// Storage disk is empty, perhaps ignored disk or not available.
+				return errDiskNotFound
+			}
+			info, err := disks[index].DiskInfo()
+			if err != nil {
+				if IsErr(err, baseErrs...) {
+					return err
+				}
+				reqInfo := (&logger.ReqInfo{}).AppendTags("disk", disks[index].String())
+				ctx := logger.SetReqInfo(context.Background(), reqInfo)
+				logger.LogIf(ctx, err)
+			}
+			disksInfo[index] = info
+			return nil
+		}, index)
 	}
-	encoded, err := e.encoder.Split(data)
-	if err != nil {
-		logger.LogIf(ctx, err)
-		return nil, err
+
+	getPeerAddress := func(diskPath string) (string, error) {
+		hostPort := strings.Split(diskPath, SlashSeparator)[0]
+		// Host will be empty for xl/fs disk paths.
+		if hostPort == "" {
+			return "", nil
+		}
+		thisAddr, err := xnet.ParseHost(hostPort)
+		if err != nil {
+			return "", err
+		}
+		return thisAddr.String(), nil
 	}
-	if err = e.encoder.Encode(encoded); err != nil {
-		logger.LogIf(ctx, err)
-		return nil, err
+
+	onlineDisks = make(madmin.BackendDisks)
+	offlineDisks = make(madmin.BackendDisks)
+	// Wait for the routines.
+	for i, err := range g.Wait() {
+		peerAddr, pErr := getPeerAddress(disksInfo[i].RelativePath)
+		if pErr != nil {
+			continue
+		}
+		if _, ok := offlineDisks[peerAddr]; !ok {
```

Comment:
Less code, but maybe too clever:

```Go
		offlineDisks[peerAddr]++
		onlineDisks[peerAddr]++
		if err == nil {
			offlineDisks[peerAddr]--
		} else {
			onlineDisks[peerAddr]--
		}
```

Reply:
Will keep the current style, this may not be well understood.